sw/airborne/modules/flight_plan_guided/flight_plan_guided.c

```diff
@@ -0,0 +1,352 @@
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  if (descent_rate > 0) {
+    guidance_v_set_guided_vz(descent_rate);
+    guidance_h_set_guided_body_vel(0, 0);
+  } else { return true; }
+
+  return false;
+}
+
+/**
+ * Hover: MAV will hover on spot in guided mode
+ * @param[in] alt, altitude at where the hover should be
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+uint8_t Hover(float alt)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+  // Horizontal velocities are set to zero and heading should be fixed
+  guidance_h_set_guided_heading(stateGetNedToBodyEulers_f()->psi);
+  guidance_h_set_guided_body_vel(0, 0);
+  guidance_v_set_guided_z(-alt);
+
+  return false;
+}
+
+
+
+/********************************************//**
+ *  Rotating functions
+ ***********************************************/
+float wanted_heading = 0;
+/**
+ * RotateToHeading: Rotate to a given heading in guided mode
+ * @param[in] heading, rotate to the wanted heading in [rad] (clockwise is positive)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool RotateToHeading(float heading)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+
+  guidance_h_set_guided_heading(heading);
+  return false;
+}
+
+/**
+ * SpeedRotateToHeading: Will slowly turn to a given heading with a fixed speed
+ * @param[in] rotation_rate, speed to slowly rotate the heading(clockwise is positive)
+ * @param[in] heading, here the MAV will stop spinning [rad] (clockwise is positive)
+ *  @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool SpeedRotateToHeading(float rotation_rate, float heading)
+{
+  guidance_h_set_guided_heading_rate(rotation_rate);
+
+  if (fabs(wrapToPi_guided(heading) - stateGetNedToBodyEulers_f()->psi) < 0.1) {
+    guidance_h_set_guided_heading(heading);
+    return false;
+  }
+  return true;
+}
+
+/********************************************//**
+ *  Blocking conditions
+ ***********************************************/
+
+/**
+ * WaitUntilAltitude: Blocks while waiting until a certain altitude is reached
+ * @param[in] altitude in [m] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilAltitude(float altitude)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+  if (stateGetPositionEnu_f()->z < altitude) { return true; }
+
+  return false;
+}
+
+
+/**
+ * WaitUntilAltitude: Blocks while waiting until a certain altitude is reached
+ * @param[in] altitude in [m] (positive z-axis is up)
+ * @param[return] Boolean for flight plan to keep running (true) or to just run once (false)
+ */
+bool WaitUntilGround(void)
+{
+  if (autopilot.mode != AP_MODE_GUIDED) { return true; }
+  if (stateGetPositionEnu_f()->z > 0.1) { return true; }
```

Comment:
And what if ground is not the `z=0` plane ? Why not have a generic `WaitUntilAlt` that looks for a range [alt-epsilon, a+epsilon]. Or if you really need, and extra option to tell if you wait until below, over or equal to a given altitude.